scraper/scraper.ts

```diff
@@ -1,28 +1,974 @@
+    const link = element.match(myRe)
+    if (link !== null && link.length > 0) {
+      if (regex.test(link[1])) {
+        const url = base + link[1]
+        urlSet.add(url)
+      }
+    }
+  })
+
+  return Array.from(urlSet)
+}
+
+/**
+ * Breaks the page down into relevant chunks from which data can be extracted
+ * @param page: page to be broken down into chunks
+ */
+const getChunks = async (page: puppeteer.Page): Promise<PageData[]> => {
+  const chunks: PageData[] = await page.evaluate(() => {
+    const chunkList: HTMLElement[][] = []
+    const courseTables: NodeListOf<HTMLElement> = document.querySelectorAll(
+      '[class="formBody"][colspan="3"]'
+    )
+    for (const course of courseTables) {
+      // Get every td which has more than 1 table
+      const parts = Array.from(course.children).filter(
+        (element: HTMLElement) => element.tagName === 'TABLE'
+      ) as HTMLElement[]
+      if (parts.length > 1) {
+        chunkList.push(parts)
+      }
+    }
+
+    // Data to be returned ----> list of course objects
+    const coursesDataElements: PageData[] = []
+
+    // Deciding...
+    for (const course of chunkList) {
+      let course_info: Chunk
+      let classes: Chunk[]
+      for (const chunk of course) {
+        // If there are any data fields inside the chunk, then categorize it
+        const data: HTMLElement = chunk.querySelector('.data')
+        if (data) {
+          const query: string = 'a[href="#top"]'
+          const classlist: HTMLElement = chunk.querySelector(query)
+          const note: HTMLElement = chunk.querySelector('.note')
+          const subtables: NodeListOf<HTMLElement> = chunk.querySelectorAll(
+            ':scope > tbody > tr > td > table'
+          )
+
+          // If the table has any element with id top, then it is the classes table.
+          if (classlist) {
+            // The table represents a classlist!! -> the split chunks are then each element of subtables...
+            // get all elements with class data from each table and then return that array.
+            const tablelist: Chunk[] = []
+            for (const subtable of subtables) {
+              tablelist.push(
+                [...subtable.querySelectorAll<HTMLElement>('.data')].map(
+                  element => element.innerText
+                )
+              )
+            }
+            if (!classes) {
+              classes = tablelist
+            } else {
+              classes = classes.concat(tablelist)
+            }
+          } else if (note) {
+            // The table is the summary table ---> skip!
+          } else if (subtables.length === 3) {
+            // This should be the course info table. --> get data elements
+            course_info = [...chunk.querySelectorAll<HTMLElement>('.data')].map(
+              element => element.innerText
+            )
+          }
+          // Else -> other heading tables ---> skip!
+        }
+      }
+      if (classes) {
```

Comment:
Following the above suggestion
```suggestion
      if (classes.length) {
```

Reply:
Updated